## Introduction to AgentChat 
We’ll use the AgentChat API to create a simple `AssistantAgent` and explore its capabilities.  
`AssistantAgent` is a built-in agent that uses a language model and has the ability to use tools.

In [3]:
# Import necessary modules
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_agentchat.messages import UserMessage
from dotenv import load_dotenv
import os

load_dotenv()
api_key = os.getenv('GEMINI_API_KEY')

In [ ]:
# Initialize the google model client
model_client = OpenAIChatCompletionClient(model = 'gemini-2.0-flash',
                                          api_key = api_key)

In [ ]:
# let's test the model client 
response = await model_client.create([UserMessage(content="What is the capital of France?", 
                                                  source="user")])
print(type(response)) 
print(response) # This will print the response object
print(response.content) # This will print the content of the response
await model_client.close() # Close the model client connection
# Note: The model client should be closed after use to free up resources (Optional)

<class 'autogen_core.models._types.CreateResult'>
finish_reason='stop' content='The capital of France is **Paris**.\n' usage=RequestUsage(prompt_tokens=7, completion_tokens=9) cached=False logprobs=None thought=None
The capital of France is **Paris**.



## Creating First AssistantAgent  
The `AssistantAgent` is a versatile agent for conversations, powered by AgentChat.  

In [11]:
# defining the AssistantAgent
model_client = OpenAIChatCompletionClient(model = 'gemini-2.0-flash',
                                          api_key = api_key)
# Creating First AssistantAgent
assistant = AssistantAgent(name = "assistant",
                           model_client = model_client, 
                           description = 'Assistant Agent for general conversation',
                           system_message = "You are a helpful assistant.")

## Testing the AssistantAgent  
We’ll use the `run` method to send a task and get a response.  

In [12]:
result = await assistant.run(task="What's the capital of USA & write 1 line about the same?")

print(result)

messages=[TextMessage(id='1c63b7d5-2da2-45b5-b297-2c260a004788', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 8, 4, 16, 10, 21, 148981, tzinfo=datetime.timezone.utc), content="What's the capital of USA & write 1 line about the same?", type='TextMessage'), TextMessage(id='afb336e6-060c-417f-b0ff-aade57f5d928', source='assistant', models_usage=RequestUsage(prompt_tokens=22, completion_tokens=27), metadata={}, created_at=datetime.datetime(2025, 8, 4, 16, 10, 22, 478407, tzinfo=datetime.timezone.utc), content="The capital of the USA is Washington, D.C. It's where the U.S. federal government is located.\n", type='TextMessage')] stop_reason=None


messages=[TextMessage(id='1c63b7d5-2da2-45b5-b297-2c260a004788', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 8, 4, 16, 10, 21, 148981, tzinfo=datetime.timezone.utc), 

content="What's the capital of USA & write 1 line about the same?", type='TextMessage'), 

TextMessage(id='afb336e6-060c-417f-b0ff-aade57f5d928', source='assistant', 

models_usage=RequestUsage(prompt_tokens=22, completion_tokens=27), metadata={}, 

created_at=datetime.datetime(2025, 8, 4, 16, 10, 22, 478407, tzinfo=datetime.timezone.utc), 

content="The capital of the USA is Washington, D.C. It's where the U.S. federal government is located.\n", type='TextMessage')] stop_reason=None


In [18]:
print(result.messages)

[TextMessage(id='1c63b7d5-2da2-45b5-b297-2c260a004788', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 8, 4, 16, 10, 21, 148981, tzinfo=datetime.timezone.utc), content="What's the capital of USA & write 1 line about the same?", type='TextMessage'), TextMessage(id='afb336e6-060c-417f-b0ff-aade57f5d928', source='assistant', models_usage=RequestUsage(prompt_tokens=22, completion_tokens=27), metadata={}, created_at=datetime.datetime(2025, 8, 4, 16, 10, 22, 478407, tzinfo=datetime.timezone.utc), content="The capital of the USA is Washington, D.C. It's where the U.S. federal government is located.\n", type='TextMessage')]


In [ ]:
print(result.messages[0].content)  # This will print the content of the first response
# Note: The messages attribute contains the conversation history, including the response from the assistant.

What's the capital of USA & write 1 line about the same?


In [ ]:
print(result.messages[-1].content) # This will print the content of the last response
# Note: The messages attribute contains the conversation history, including the response from the assistant.

The capital of the USA is Washington, D.C. It's where the U.S. federal government is located.



## Streaming Messages
We can also stream each message as it is generated by the agent by using the `run_stream()` method, and use Console to print the messages as they appear to the `console`.


In [19]:
from autogen_agentchat.ui import Console

async def assistant_run_stream() -> None:
    # Option 1: read each message from the stream (as shown in the previous example).
    # async for message in agent.run_stream(task="Find information on AutoGen"):
    #     print(message)

    # Option 2: use Console to print all messages as they appear.
    await Console(
        assistant.run_stream(task="write something in 2-3 lines on AutoGen!"),
        output_stats=True,  # Enable stats printing.
    )


# Use asyncio.run(assistant_run_stream()) when running in a script.
await assistant_run_stream()

---------- TextMessage (user) ----------
write something in 2-3 lines on AutoGen!
---------- TextMessage (assistant) ----------
AutoGen is a framework developed by Microsoft that enables building AI applications with multiple agents that can converse with each other to solve tasks. These agents can be customized with different roles, skills, and access to tools, fostering collaborative problem-solving in a dynamic and automated manner. It simplifies the creation of complex AI workflows by orchestrating interactions between different specialized AI agents.

[Prompt tokens: 61, Completion tokens: 73]
---------- Summary ----------
Number of messages: 2
Finish reason: None
Total prompt tokens: 61
Total completion tokens: 73
Duration: 1.69 seconds


In [ ]:
# Integration with ollama
# first pip install below packages
#pip install -U "autogen-ext[ollama]"
# from autogen_core.models import UserMessage
# from autogen_ext.models.ollama import OllamaChatCompletionClient

# # Assuming your Ollama server is running locally on port 11434.
# ollama_model_client = OllamaChatCompletionClient(model="llama3.2")

# response = await ollama_model_client.create([UserMessage(content="What is the capital of France?", source="user")])
# print(response)
# await ollama_model_client.close()